src/warp10-datasource.ts

```diff
@@ -260,24 +282,50 @@ export default class Warp10Datasource {
         value = value.replace(/'/g, '"')
       if (typeof value === 'string' && !value.startsWith('<%') && !value.endsWith('%>'))
         value = `'${value}'`
-      wsHeader += `${value || 'NULL'} '${myVar}' STORE `
+      wsHeader += `${value || 'NULL'} '${myVar}' STORE\n`
     }
     // Dashboad templating vars
+    // current.text is the label. In case of multivalue, it is a string 'valueA + valueB'
+    // current.value is a string, depending on query output. In case of multivalue, it is an array of strings. array contains "$__all" if user selects All.
+    console.log("this.templateSrv.variables", this.templateSrv.variables)
+
     for (let myVar of this.templateSrv.variables) {
-      let value = myVar.current.text
-
-      if (myVar.current.value != null && (myVar.current.value === '$__all' || (myVar.current.value.length === 1 && myVar.current.value[0] === '$__all')))
-      {
-        if (myVar.allValue !== null)
-          value = myVar.allValue;
-        else
-          value = myVar.options.slice(1).map(e => e.text).join(" + ");
+      let value = myVar.current.value;
+
+      if (Array.isArray(value) && (value.length == 1 && value[0] === '$__all')) {
+        // User checked the "select all" checkbox
+        if (myVar.allValue && myVar.allValue !== "") {
+          // User also defined a custom value in the variable settings
+          let customValue: String = myVar.allValue;
+          wsHeader += `[ '${customValue}' ] '${myVar.name}' STORE\n`
+          // If user already starts its custom value with ~, remove it.
+          if (customValue.startsWith("~")) {
+            customValue = customValue.slice(1);
+          }
+          wsHeader += ` '~${customValue}' '${myVar.name}_wsregexp' STORE\n`
+        } else {
+          // if no custom all value is defined :
+          // it means we shall create a list of all the values in WarpScript from options, ignoring "$__all" special option value. 
+          let allValues: String[] = myVar.options.filter(o => o.value !== "$__all").map(o => o.value);
+          wsHeader += `[ ${allValues.map(s => `'${s}'`).join(" ")} ] '${myVar.name}' STORE\n`; // all is stored as string in generated WarpScript.
+          //also create a ready to use regexp, suffixed by _wsregexp
+          wsHeader += ` '~' $${myVar.name} REOPTALT + '${myVar.name}_wsregexp' STORE\n`
+        }
+      } else if (Array.isArray(value)) {
```

Comment:
Depending on Grafana version, array might contain only 1 value.

Reply:
With or without multi selection enabled ?

Reply:
Did not checked and do not have 5.X anymore to test. Anyway, It is safer to assume "1 value + multi == single value" and also ensure better user understanding (IMO).

Reply:
done like this.